ledger/streaming/client/src/ledgerdb_sink.rs

```diff
@@ -0,0 +1,115 @@
+    stream: BS,
+
+    /// Message channel that sends data to the db
+    sender: mpsc::Sender<BlockData>,
+
+    /// Logger running
+    logger: Logger,
+}
+
+impl<BS: Stream<Item = StreamResult<BlockData>> + std::marker::Unpin> BlockSink<BS> {
+    pub fn new(stream: BS, logger: Logger, sender: mpsc::Sender<BlockData>) -> Self {
+        Self {
+            stream,
+            logger,
+            sender,
+        }
+    }
+
+    // Ingest blocks from stream into a sink for processing
+    async fn broadcast(&mut self) {
+        while let Some(block) = self.stream.next().await {
+            if let Ok(block) = block {
+                let index = block.block().index;
+                match self.sender.send(block).await {
+                    Ok(_) => log::debug!(self.logger, "successfully sent block {}", index),
+                    Err(_) => log::warn!(self.logger, "receiver failed to send"),
+                }
+            }
+        }
+    }
+}
+
+/// Write to the ledger in a synchronous loop that consistently polls for blocks
+pub fn sink(
+    ledger_ref: Arc<Mutex<impl Ledger>>,
+    mut receiver: mpsc::Receiver<BlockData>,
+    logger: Logger,
+) {
+    let mut ledger = ledger_ref.lock().unwrap();
+    loop {
+        if let Some(block_data) = receiver.blocking_recv() {
+            let index = block_data.block().index;
+            ledger
+                .append_block(
+                    block_data.block(),
+                    block_data.contents(),
+                    block_data.signature().clone(),
+                )
+                .unwrap();
+            log::trace!(logger, "Block {} written", index)
+        } else {
+            log::warn!(logger, "Receiver shut down!");
+            break;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mc_common::logger::{test_with_logger, Logger};
+    use mc_ledger_db::{test_utils::get_mock_ledger, Ledger};
+    use std::sync::{Arc, Mutex};
+    use tokio::runtime::Builder;
+    #[test_with_logger]
+    fn assert_blocks_arrive(logger: Logger) {
+        let source_ledger = get_mock_ledger(200);
+        let mut test_block_data: Vec<StreamResult<BlockData>> = Vec::new();
+        let (tx, rcv) = mpsc::channel(10000);
+
+        for i in 1..200 {
+            test_block_data.push(Ok(source_ledger.get_block_data(i).unwrap()));
+        }
+        let stream = futures::stream::iter(test_block_data);
+        let mut bs = BlockSink::new(stream, logger.clone(), tx);
+
+        let dest_ledger = Arc::new(Mutex::new(get_mock_ledger(1)));
+        let ledger_ref = dest_ledger.clone();
+        std::thread::spawn(move || sink(ledger_ref, rcv, logger.clone()));
+
+        let rt = Builder::new_current_thread().enable_all().build().unwrap();
+
+        rt.block_on(async move {
+            let mut counter = 0usize;
+            loop {
+                counter += 1;
+                bs.broadcast().await;
+                if counter < 200 {
```

Comment:
Did you mean `>` here? As written, this breaks after the first iteration.

Reply:
Indeed I did!

Reply:
Indeed I Did!